### PR TITLE
feature: use patch strategy to update services weight

### PR DESCRIPTION
**Pull request by yyzxw**
**What this PR does / why we need it**:

use patch strategy to update services weight to avoid strongly dependent versions of contour.

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by wilsonwu** [comment]
Thanks for this good feature, @izturn please help to review.

File: pkg/plugin/plugin.go [hunk 1]
```diff
@@ -70,10 +68,10 @@ func (r *RpcPlugin) SetWeight(rollout *v1alpha1.Rollout, desiredWeight int32, ad
 	ctx := context.Background()
 
 	for _, proxy := range ctr.HTTPProxies {
-		slog.Debug("updating httpproxy", slog.String("name", proxy))
+		slog.Debug("updating httpProxy weight", slog.String("name", proxy))
```
Comment (izturn) [inline pkg/plugin/plugin.go:71]: Unify the all name to `httpProxy` or `httpproxy` ?

File: pkg/plugin/plugin.go [hunk 2]
```diff
@@ -266,7 +278,7 @@ func getServiceMap(httpProxy *contourv1.HTTPProxy) map[string]*contourv1.Service
 
 func validateRolloutParameters(rollout *v1alpha1.Rollout) error {
 	if rollout == nil || rollout.Spec.Strategy.Canary == nil || rollout.Spec.Strategy.Canary.StableService == "" || rollout.Spec.Strategy.Canary.CanaryService == "" {
-		return fmt.Errorf("illegal parameter(s)")
+		return fmt.Errorf("illegal parameter(s), canary service and stable service must be specified")
```
Comment (izturn) [inline pkg/plugin/plugin.go:281]: both canary and stable service must be specified.

File: pkg/plugin/plugin.go [hunk 3]
```diff
@@ -180,6 +167,31 @@ func (r *RpcPlugin) updateHTTPProxy(ctx context.Context, httpProxyName string, r
 	return nil
 }
 
+func createPatch(httpProxy *contourv1.HTTPProxy, rollout *v1alpha1.Rollout, desiredWeight int32) ([]byte, types.PatchType, error) {
+	oldData, err := json.Marshal(httpProxy.DeepCopy())
+	if err != nil {
+		return nil, "", fmt.Errorf("serializing current configuration: %w", err)
```
Comment (izturn) [inline pkg/plugin/plugin.go:173]: failed to marshal the current configuration: %w

File: pkg/plugin/plugin.go [hunk 4]
```diff
@@ -180,6 +167,31 @@ func (r *RpcPlugin) updateHTTPProxy(ctx context.Context, httpProxyName string, r
...
+		return nil, "", err
+	}
+
+	canarySvc.Weight = int64(desiredWeight)
+	stableSvc.Weight = 100 - canarySvc.Weight
+	slog.Debug("old weight", slog.Int64("canary", canarySvc.Weight), slog.Int64("stable", stableSvc.Weight))
+
+	newData, err := json.Marshal(httpProxy)
+	if err != nil {
+		return nil, "", fmt.Errorf("serializing current configuration: %w", err)
```
Comment (izturn) [inline pkg/plugin/plugin.go:187]: as above

File: pkg/plugin/plugin.go [hunk 5]
```diff
@@ -180,6 +167,31 @@ func (r *RpcPlugin) updateHTTPProxy(ctx context.Context, httpProxyName string, r
...
+	}
+
+	// now default use json merge patch.
```
Comment (izturn) [inline pkg/plugin/plugin.go:190]: remove this comment?

File: pkg/plugin/plugin.go [hunk 6]
```diff
@@ -148,25 +146,14 @@ func (r *RpcPlugin) updateHTTPProxy(ctx context.Context, httpProxyName string, r
...
-
-	m, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&httpProxy)
-	if err != nil {
-		return err
-	}
-	updated, err := r.dynamicClient.Resource(contourv1.HTTPProxyGVR).Namespace(rollout.Namespace).Update(ctx, &unstructured.Unstructured{Object: m}, metav1.UpdateOptions{})
+	updated, err := r.dynamicClient.Resource(contourv1.HTTPProxyGVR).Namespace(rollout.Namespace).Patch(ctx, httpProxyName, patchType, patchData, metav1.PatchOptions{})
 	if err != nil {
-		return err
+		return fmt.Errorf("failed update httpProxy: %w", err)
```
Comment (izturn) [inline pkg/plugin/plugin.go:156]: failed to  update the ....

File: pkg/plugin/plugin.go [hunk 7]
```diff
@@ -132,12 +130,12 @@ func (r *RpcPlugin) Type() string {
 func (r *RpcPlugin) getHTTPProxy(ctx context.Context, namespace string, name string) (*contourv1.HTTPProxy, error) {
 	unstr, err := r.dynamicClient.Resource(contourv1.HTTPProxyGVR).Namespace(namespace).Get(ctx, name, metav1.GetOptions{})
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get httpproxy: %w", err)
```
Comment (izturn) [inline pkg/plugin/plugin.go:133]: get -> get the ?

File: pkg/plugin/plugin.go [hunk 8]
```diff
@@ -132,12 +130,12 @@ func (r *RpcPlugin) Type() string {
...
 	}
 
 	var httpProxy contourv1.HTTPProxy
 	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstr.UnstructuredContent(), &httpProxy); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to convert httpproxy: %w", err)
```
Comment (izturn) [inline pkg/plugin/plugin.go:138]: convert -> convert the

**Comment by izturn** [comment]
@wilsonwu PTAL

**Comment by izturn** [comment]
@yyzxw LGTM, thx

**Review by wilsonwu — APPROVED** [review]
LGTM, thanks!

**Pull request merged**